src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs

```diff
@@ -1255,64 +1255,66 @@ .locals init (int V_0,
-            //       IL_0007:  dup
-            //       IL_0008:  ldind.i4
-            //       IL_0009:  ldc.i4.2
-            //       IL_000a:  add
-            //       IL_000b:  dup
-            //       IL_000c:  stloc.3
-            //       IL_000d:  stind.i4
-            //       IL_000e:  ldloc.3
-            //       IL_000f:  ldarg.0
-            //       IL_0010:  ldc.i4.3
-            //       IL_0011:  ldelema    "int"
-            //       IL_0016:  stloc.0
-            //       IL_0017:  ldloc.0
-            //       IL_0018:  ldind.i4
-            //       IL_0019:  stloc.1
-            //       IL_001a:  call       "System.Threading.Tasks.Task<int> Test.G()"
-            //       IL_001f:  call       "int System.Runtime.CompilerServices.AsyncHelpers.Await<int>(System.Threading.Tasks.Task<int>)"
-            //       IL_0024:  stloc.2
-            //       IL_0025:  ldloc.0
-            //       IL_0026:  ldloc.1
-            //       IL_0027:  ldloc.2
-            //       IL_0028:  add
-            //       IL_0029:  dup
-            //       IL_002a:  stloc.3
-            //       IL_002b:  stind.i4
-            //       IL_002c:  ldloc.3
-            //       IL_002d:  ldc.i4.4
-            //       IL_002e:  call       "int Test.H(int, int, int)"
-            //       IL_0033:  pop
-            //       IL_0034:  ldc.i4.1
-            //       IL_0035:  ret
-            //     }
-            //     """);
+            var verifier = CompileAndVerify(comp, verify: Verification.Fails with
+            {
+                ILVerifyMessage = """
+                    [F]: Unexpected type on the stack. { Offset = 0x40, Found = Int32, Expected = ref '[System.Runtime]System.Threading.Tasks.Task`1<int32>' }
+                    """
+            });
+
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("Test.F(int[])", """
+                {
+                  // Code size       65 (0x41)
+                  .maxstack  4
+                  .locals init (int[] V_0,
+                                int& V_1,
+                                int V_2,
+                                int V_3,
+                                int V_4)
+                  IL_0000:  ldarg.0
+                  IL_0001:  ldc.i4.1
+                  IL_0002:  ldelema    "int"
+                  IL_0007:  dup
+                  IL_0008:  ldind.i4
+                  IL_0009:  ldc.i4.2
+                  IL_000a:  add
+                  IL_000b:  dup
+                  IL_000c:  stloc.s    V_4
+                  IL_000e:  stind.i4
+                  IL_000f:  ldloc.s    V_4
+                  IL_0011:  ldarg.0
+                  IL_0012:  stloc.0
+                  IL_0013:  ldloc.0
+                  IL_0014:  ldc.i4.3
+                  IL_0015:  ldelem.i4
+                  IL_0016:  pop
+                  IL_0017:  ldloc.0
+                  IL_0018:  ldc.i4.3
+                  IL_0019:  ldelem.i4
+                  IL_001a:  stloc.2
```

Comment:
This seems to load `array[3]` twice, discarding the first value. Unexpected?

Reply:
> Thanks. FWIW, it looks like the state machine IL above also had this problem (`IL_003d: pop`)

Thanks for looking at that Julien. Given that, I plan to leave this as is for now, unless it is fairly obvious from a quick investigation.

Reply:
I think the extra/discarded read is the sacrificial evaluation that we're discussing elsewhere.
Is the following correct:
When doing an assignment `arr[index] = await M();`, we cache parts with side-effects, and do a sacrificial evaluation of the hoisted/stripped indexing operation *before* the await. The index-out-of-bounds exception is thrown *before* `M()` is evaluated.

Then I think what we're seeing here is that in a compound assignment `arr[index] += await M();` we're doing a sacrificial evaluation and an indexing/read before the await, then an indexing/write after the await.